wagtail/documents/rich_text.py

```diff
@@ -18,10 +35,49 @@ def expand_db_attributes(attrs):
             return "<a>"
 
 
-def document_linktype_handler(attrs):
-    Document = get_document_model()
-    try:
-        doc = Document.objects.get(id=attrs['id'])
-        return '<a href="%s">' % escape(doc.url)
-    except Document.DoesNotExist:
-        return "<a>"
+EditorHTMLDocumentLinkConversionRule = [
+    editor_html.LinkTypeRule('document', DocumentLinkHandler),
+]
+
+
+# draft.js / contentstate conversion
+
+def document_link_entity(props):
+    """
+    Helper to construct elements of the form
+    <a id="1" linktype="document">document link</a>
+    when converting from contentstate data
+    """
+
+    return DOM.create_element('a', {
+        'linktype': 'document',
+        'id': props.get('id'),
+    }, props['children'])
+
+
+class DocumentLinkElementHandler(LinkElementHandler):
+    """
+    Rule for populating the attributes of a document link when converting from database representation
+    to contentstate
+    """
+    def get_attribute_data(self, attrs):
+        Document = get_document_model()
+        try:
+            doc = Document.objects.get(id=attrs['id'])
+        except Document.DoesNotExist:
+            return {}
+
+        return {
+            'id': doc.id,
+            'url': doc.url,
+        }
+
+
+ContentstateDocumentLinkConversionRule = {
+    'from_database_format': {
+        'a[linktype="document"]': DocumentLinkElementHandler('DOCUMENT'),
```

Comment:
`'DOCUMENT'` should be `ENTITY_TYPES.DOCUMENT`, or `ENTITY_TYPES.DOCUMENT` below should be `'DOCUMENT'`.

Reply:
This is resolved as far as I can see.